feat: enhance height handling, add view argument to open()/toggle()

drowning-cat opened the pull request:
1. ~~Add new options~~

```lua
-- Restore the previous quickfix window view parameters, such as height and cursor position, on `quicker.open` and `quicker.toggle`
winrestore = {
  -- Restore the height of the quickfix window
  height = false,
  -- Respect the min_height and max_height arguments for `quicker.open` and `quicker.toggle`
  height_minmax = false,
  -- Restore cursor and scroll positions. See `:h winsaveview()`
  view = true,
},
```

2. Prevent the scroll position from jumping when the quickfix opens

https://github.com/user-attachments/assets/7ce9350d-33ba-4cb7-949f-6d267f34cd97

drowning-cat commented:
> [!WARNING]
> There is also a bug to fix: the open/toggle is unusable with vertical split because the `min_height` and `max_height` properties are incompatible with it.

```lua
require('quicker').toggle { open_cmd_mods = { vertical = true } }
```

stevearc reviewed (changes requested):
This kind of config and feature is likely to grow out of control in the future. It starts here, but then someone will ask for a way to restore the height for only certain quickfixes, or they want to clear the stored height/view when the quickfix changes, or they _don't_ want to clear the height/view when the quickfix changes. I would much rather look at how the user can customize this in their own config, and then see if there are any primitives we can expose to make that simpler/easier.

Right now I think most of this should be easy to do in end-user config. Just set up a `WinClosed` autocmd, check if the buffer is quickfix, and store the height/view. The `toggle` and `open` commands already let you specify a height and min/max height, so that's pretty easy. The only thing that we could maybe do is make it slightly easier to restore the saved win view, and honestly maybe just returning the winid from `open` and `toggle` would be enough. What do you think?

drowning-cat commented:
What you’re suggesting does make sense, but there are quite a lot of configurations required to implement saving the state for each window (give me a few more days).

For now, I’d like to point out a few things.

#### 1. Remove `min_height` and `max_height` (optional, breaking change)

I suggest removing `min_height` and `max_height` since they don’t align with the command interface (`:copen`, `:lopen`, etc.) and must be set along with `height`.

Instead of this:

```lua
require('quicker').toggle { height = 10, max_height = 10, min_height = 10 }
```

I’d prefer something like this:

```lua
require('quicker').toggle { focus = true, height = 10 }
```

```lua
-- local clamp = require('quicker.util').clamp
local clamp = function(min, val, max)
  return math.max(min, math.min(max, val))
end

require('quicker').toggle {
  focus = true,
  height = clamp(4, #vim.fn.getqflist(), 10)
}
```

By default, `height` should be set using:

```lua
{
  height = clamp(4, number_of_entries, 10)
}
```

<details><summary>Or perhaps, as an alternative interface, `height` could be implemented as a function</summary>

```lua
height = function(entries)
  return require('quicker.util').clamp(4, #entries, 10)
end
```

</details>

#### 2. Add `width` option (reasonable)

I’d also recommend adding a `width` option that behaves similarly to `height` but should be ignored if not set (no default value).

For example:

```lua
require('quicker').toggle {
  -- height is ignored by default
  width = 20,
  open_cmd_mods = { vertical = true }
}
```

#### 3. Manage view restoration via plugin (pending)

I think view restoration should be handled by the plugin itself, restoring the view for each window (not like my current implementation, which is stupid).

This should be the default behavior because it’s much easier to press `g` or `gg` to navigate within the quickfix list than to manually find a specific position.

As a side effect, the plugin could:
 - expose some utility functions for this behavior
 - provide examples in the source code on how to achieve it

There’s no rush—I just want to recreate this from a user’s perspective and will paste a solution here for further discussion.
Maybe it's not as scary as I think.

drowning-cat commented:
```lua
---@alias WinSaveDict { height: integer, view: vim.fn.winsaveview.ret }

---@type table<integer, WinSaveDict>
local win_config = {}

local get_qfid = function(winid)
  local wininfo = vim.fn.getwininfo(winid)[1]
  if wininfo.loclist == 1 then
    return vim.fn.getloclist(winid, { id = 0 }).id
  end
  if wininfo.quickfix == 1 then
    return vim.fn.getqflist({ id = 0 }).id
  end
end

local save_qf_window = function()
  local winid = vim.fn.win_getid()
  local qfid = get_qfid(winid)
  win_config[qfid] = {
    height = vim.api.nvim_win_get_height(winid),
    view = vim.fn.winsaveview(),
  }
  if qfid ~= 0 then
    table.remove(win_config, 0)
  end
end
local restore_qf_window = function()
  local winid = vim.fn.win_getid()
  local qfid = get_qfid(winid)
  if win_config[qfid] then
    vim.api.nvim_win_set_height(winid, win_config[qfid].height)
    vim.fn.winrestview(win_config[qfid].view)
  end
end

-- :close
vim.api.nvim_create_autocmd('WinClosed', {
  callback = function()
    if vim.bo.ft == 'qf' then
      save_qf_window()
    end
  end,
})

local prevent_qf_restore_schedule = false

-- :copen; :cnewer, :colder
vim.api.nvim_create_autocmd('BufReadPost', {
  pattern = 'quickfix',
  callback = function()
    vim.schedule(function()
      if not prevent_qf_restore_schedule then
        restore_qf_window()
      end
      prevent_qf_restore_schedule = false
    end)
  end,
})

--- @param opts quicker.OpenOpts
local quicker_toggle = function(opts)
  local qfwinid = vim.fn.getqflist({ winid = 0 }).winid
  local qfid = vim.fn.getqflist({ id = 0 }).id

  local height = (win_config[qfid] or {}).height or 10

  -- height = clamp(4, height, 16)

  opts.min_height = opts.height or height
  opts.max_height = opts.height or height
  opts.height = opts.height or height

  require('quicker').toggle(opts)

  local is_open = qfwinid ~= 0
  local view = (win_config[qfid] or {}).view

  if is_open and view then
    if not opts.focus then
      local goback_winid = vim.fn.win_getid()
      vim.fn.win_getid(qfwinid)
      vim.fn.winrestview(view)
      vim.fn.win_gotoid(goback_winid)
    else
      vim.fn.winrestview(view)
    end
  end

  prevent_qf_restore_schedule = true
end

return {
  {
    'stevearc/quicker.nvim',
    ft = 'qf',
    opts = {},
    keys = {
      {
        '<leader>q',
        function()
          quicker_toggle { focus = true }
        end,
        { desc = 'Toggle [q]quickfix list' },
      },
      {
        '<leader>l',
        function()
          quicker_toggle { loclist = true, focus = true }
        end,
        { desc = 'Toggle [l]oclist list' },
      },
    },
  },
}
```

<details><summary>With <i>commentaries</i></summary>

```lua
---@alias WinSaveDict { height: integer, view: vim.fn.winsaveview.ret }

---@type table<integer, WinSaveDict>
-- I believe a table like { 0 = {}, 1000 = {}, 25 =  {} } will not cause any trouble
local win_config = {}

local get_qfid = function(winid)
  local wininfo = vim.fn.getwininfo(winid)[1]
  -- `getloclist` and `getqflist` both return 0 if no list is defined. However, you cannot open an empty location list.
  -- This means the ID is 0 only for an empty quickfix list.
  -- We may clear it after the first defined quickfix list.

  -- Order of checks does matter.
  -- For the location list, wininfo.loclist = 1 and wininfo.quickfix = 1
  if wininfo.loclist == 1 then
    return vim.fn.getloclist(winid, { id = 0 }).id
  end
  if wininfo.quickfix == 1 then
    return vim.fn.getqflist({ id = 0 }).id
  end
end

local save_qf_window = function()
  local winid = vim.fn.win_getid()
  local qfid = get_qfid(winid)
  -- Quickfix and location list IDs do not overlap.
  -- Yay, we can simplify the code.
  win_config[qfid] = {
    height = vim.api.nvim_win_get_height(winid),
    view = vim.fn.winsaveview(),
  }
  -- Optional. Remove config after defining a non-empty quickfix list
  if qfid ~= 0 then
    table.remove(win_config, 0)
  end
end
local restore_qf_window = function()
  local winid = vim.fn.win_getid()
  local qfid = get_qfid(winid)
  if win_config[qfid] then
    vim.api.nvim_win_set_height(winid, win_config[qfid].height)
    vim.fn.winrestview(win_config[qfid].view)
  end
end

-- :close
vim.api.nvim_create_autocmd('WinClosed', {
  callback = function()
    if vim.bo.ft == 'qf' then
      save_qf_window()
    end
  end,
})

-- Optional. Prevent restoring the quickfix window twice
local prevent_qf_restore_schedule = false

-- :copen; :cnewer, :colder
vim.api.nvim_create_autocmd('BufReadPost', {
  pattern = 'quickfix',
  callback = function()
    -- Scheduling is required. I'm not sure how to eliminate it. It's causing annoying height and cursor jumps.
    -- To prevent such jumping for `quicker.toggle`, I need to write a lot of garbage code (refer to quicker_toggle).
    -- However, I’m unable to resolve the jumping issue for commands such as :copen, :cnewer, and :colder.
    vim.schedule(function()
      if not prevent_qf_restore_schedule then
        restore_qf_window()
      end
      prevent_qf_restore_schedule = false
    end)
  end,
})

--- @param opts quicker.OpenOpts
local quicker_toggle = function(opts)
  local qfwinid = vim.fn.getqflist({ winid = 0 }).winid
  local qfid = vim.fn.getqflist({ id = 0 }).id

  -- Restore height

  local height = (win_config[qfid] or {}).height or 10

  -- height = clamp(4, height, 16)

  opts.min_height = opts.height or height
  opts.max_height = opts.height or height
  opts.height = opts.height or height

  require('quicker').toggle(opts)

  -- Restore view

  local is_open = qfwinid ~= 0
  local view = (win_config[qfid] or {}).view

  if is_open and view then
    if not opts.focus then
      local goback_winid = vim.fn.win_getid()
      vim.fn.win_getid(qfwinid)
      vim.fn.winrestview(view)
      vim.fn.win_gotoid(goback_winid)
    else
      vim.fn.winrestview(view)
    end
  end

  prevent_qf_restore_schedule = true
end

return {
  {
    'stevearc/quicker.nvim',
    ft = 'qf',
    opts = {},
    keys = {
      {
        '<leader>q',
        function()
          quicker_toggle { focus = true }
        end,
        { desc = 'Toggle [q]quickfix list' },
      },
      {
        '<leader>l',
        function()
          quicker_toggle { loclist = true, focus = true }
        end,
        { desc = 'Toggle [l]oclist list' },
      },
    },
  },
}
```

</details>

drowning-cat commented:
Ok, in addition to the https://github.com/stevearc/quicker.nvim/pull/38#issuecomment-2599234615 option, I would also add a `view` option to `open/toggle`:

```lua
require('quicker').open {
  ---@type vim.fn.winrestview.dict
  view = {
    lnum = 20,
  },
}
```

I’m not sure, but I might also add a `get_qftype` utility:

```lua
local get_wininfo = function(winid)
  return vim.fn.getwininfo(winid)[1]
end

local get_qftype = function(winid)
  if get_wininfo(winid).loclist == 1 then
    return 'll' ---@type 'll'
  end
  if get_wininfo(winid).quickfix == 1 then
    return 'qf' ---@type 'qf'
  end
  return nil
end

local get_qfid = function(winid)
  if get_qftype(winid) == 'qf' then
    return vim.fn.getqflist({ id = 0 }).id
  end
  if get_qftype(winid) == 'll' then
    return vim.fn.getloclist(winid, { id = 0 }).id
  end
end
```

drowning-cat commented:
And we need to rework a little how the height property is set. It is better to set it via cmd arguments.
For example:
- `:vert botright 40 copen` - will not only open vertically, but also set the width (not height)
- `:vert botright 30 copen` - will change the size of the opened window

drowning-cat commented:
- [ ] Use `count` property to set `height` in `open/toggle`
- [ ] Pass `view` argument to `open/toggle`
- [ ] Update documentation

drowning-cat commented:
And `constrain cursor(bufnr)`, which defines 3 events, is fired on the `FileType` event, which produces a lot of events on each quickfix toggling.

drowning-cat commented:
The current implementation of `constrain_cursor()` is causing too much trouble with scheduling and needs to be refactored.

---

```lua
-- For some reason the cursor bounces back to its original position,
-- so we have to defer the call
 ```
 
I guess it's because, on the initial load, there is a layout shift with the virtual column, which causes `CursorMoved` to fire.

drowning-cat commented:
I believe https://github.com/stevearc/quicker.nvim/pull/38#issuecomment-2600971362 and https://github.com/stevearc/quicker.nvim/pull/38#issuecomment-2601014972 are not directly related to the current PR.

I guess I’m **done** working on this PR. Apologies for the lengthy comments.

stevearc reviewed (changes requested):
Current approach LGTM. Just fix the merge conflicts and we're good to go

drowning-cat commented:
I am a little bit confused because resolving conflicts (merging the master branch) resulted in an empty commit, as I messed up the git history a little bit. However, the merge conflicts have been resolved.

stevearc commented on lua/quicker/init.lua:
```diff
@@ -141,31 +144,39 @@ end
-    vim.api.nvim_win_set_height(0, height)
+  if not vim.tbl_isempty(opts.view) then
+    vim.fn.winrestview(opts.view)
   end
```
Looks like a bad merge

stevearc commented on lua/quicker/util.lua:
```diff
@@ -71,25 +71,12 @@ M.get_lnum_extmarks = function(bufnr, lnum, line_len, ns)
   end, extmarks)
 end
 
----Return true if the window is a full-height leaf window
----@param winid? integer
----@return boolean
-M.is_full_height_vsplit = function(winid)
-  if not winid or winid == 0 then
-    winid = vim.api.nvim_get_current_win()
-  end
-  local layout = vim.fn.winlayout()
-  -- If the top layout is not vsplit, then it's not a vertical leaf
-  if layout[1] ~= "row" then
-    return false
-  end
-  for _, v in ipairs(layout[2]) do
-    if v[1] == "leaf" and v[2] == winid then
-      return true
-    end
-  end
-
-  return false
```
Same here, the merge deleted this function

stevearc reviewed (changes requested):
I think there may be an issue with how you resolve merges, because if all you saw was an empty commit then that's quite wrong. The merge deleted some logic that I added to master to handle the case when opening the quickfix as a vsplit

drowning-cat commented:
I changed the merge process. Instead of "resolving" conflicts in the merged commit, I added parts of the code during the merge and then removed them with separate commits.

Sorry for wasting so much time on such a simple PR.

stevearc commented:
Great, looks good now. Thanks for the PR!
